File: runtime/python/mochi_runtime/stream.py
"""Synchronous broadcast stream primitives for the Phase 10.0 surface.

Mochi v1 streams are MPMC broadcast: each subscriber sees every value
emitted after it subscribed. Phase 10.0 supports the single-threaded
synchronous case (emit and recv interleaved on the same execution path),
mirroring the existing C-fixture corpus at tests/transpiler3/c/fixtures/
stream/. Async / cross-task broadcast is Phase 11+.

Buffer growth: each MochiStream holds an append-only list and each
MochiSub holds a read cursor. recv_sub advances the cursor by one.
Cap is recorded but not enforced (no producer-side backpressure in
the sync surface). Callers that need bounded broadcast use
asyncio.Queue under the Phase 11+ async surface.
"""

from __future__ import annotations

from typing import Generic, TypeVar

T = TypeVar("T")


class MochiStream(Generic[T]):
    __slots__ = ("_buffer", "_cap")

    def __init__(self, cap: int) -> None:
        self._buffer: list[T] = []
        self._cap = cap


class MochiSub(Generic[T]):
    __slots__ = ("_stream", "_pos")

    def __init__(self, stream: MochiStream[T]) -> None:
        self._stream = stream
        self._pos = len(stream._buffer)


def mochi_make_stream(cap: int) -> MochiStream[T]:
    return MochiStream(cap)


def mochi_subscribe(stream: MochiStream[T]) -> MochiSub[T]:
    return MochiSub(stream)


def mochi_emit(stream: MochiStream[T], val: T) -> None:
    stream._buffer.append(val)


def mochi_recv_sub(sub: MochiSub[T]) -> T:
    v = sub._stream._buffer[sub._pos]
    sub._pos += 1
    return v

File: transpiler3/python/lower/stream.go
package lower

import (
	"github.com/mochilang/mochi-python/transpiler3/c/aotir"
	"github.com/mochilang/mochi-python/transpiler3/python/pysrc"
)

// lowerStreamMakeExpr emits `mochi_make_stream(cap)`. Phase 10.0 streams
// are synchronous broadcast buffers (no producer-side backpressure); the
// `cap` argument is preserved on the MochiStream instance for future use
// but does not bound buffer growth in v1.
func (l *lowerer) lowerStreamMakeExpr(e *aotir.StreamMakeExpr) (pysrc.Expr, error) {
	l.needsStream = true
	cap, err := l.lowerExpr(e.Cap)
	if err != nil {
		return nil, err
	}
	return &pysrc.Call{
		Func: &pysrc.Name{Id: "mochi_make_stream"},
		Args: []pysrc.Expr{cap},
	}, nil
}

// lowerSubMakeExpr emits `mochi_subscribe(stream)`. The new subscriber's
// read cursor starts at the current write position; any value emitted
// before subscribe is invisible to that subscriber (broadcast semantics
// from the C fixture corpus).
func (l *lowerer) lowerSubMakeExpr(e *aotir.SubMakeExpr) (pysrc.Expr, error) {
	l.needsStream = true
	stream, err := l.lowerExpr(e.Stream)
	if err != nil {
		return nil, err
	}
	return &pysrc.Call{
		Func: &pysrc.Name{Id: "mochi_subscribe"},
		Args: []pysrc.Expr{stream},
	}, nil
}

// lowerStreamEmitStmt emits `mochi_emit(stream, val)`. Synchronous append
// to the broadcast buffer; every live subscriber observes the value on
// their next recv_sub call.
func (l *lowerer) lowerStreamEmitStmt(s *aotir.StreamEmitStmt) (pysrc.Stmt, error) {
	l.needsStream = true
	stream, err := l.lowerExpr(s.Stream)
	if err != nil {
		return nil, err
	}
	val, err := l.lowerExpr(s.Val)
	if err != nil {
		return nil, err
	}
	return &pysrc.ExprStmt{X: &pysrc.Call{
		Func: &pysrc.Name{Id: "mochi_emit"},
		Args: []pysrc.Expr{stream, val},
	}}, nil
}

// lowerSubRecvExpr emits `mochi_recv_sub(sub)`. Advances the subscriber's
// read cursor by one and returns the value at the previous cursor.
// IndexError surfaces if the caller recvs past the latest emit; v1
// fixtures never do this (every recv is preceded by a paired emit).
func (l *lowerer) lowerSubRecvExpr(e *aotir.SubRecvExpr) (pysrc.Expr, error) {
	l.needsStream = true
	sub, err := l.lowerExpr(e.Sub)
	if err != nil {
		return nil, err
	}
	return &pysrc.Call{
		Func: &pysrc.Name{Id: "mochi_recv_sub"},
		Args: []pysrc.Expr{sub},
	}, nil
}
